bmorph/core/models.py

```diff
@@ -0,0 +1,361 @@
+import os
+import copy
+import random
+import pandas as pd
+import numpy as np
+import xarray as xr
+from joblib import dump, load
+from functools import partial
+from typing import List, Tuple
+
+from sklearn.preprocessing import FunctionTransformer
+from sklearn.preprocessing import StandardScaler
+from sklearn.preprocessing import MinMaxScaler
+from sklearn.pipeline import Pipeline
+
+
+def elevation_scaler(x, feature_range=(0, 1), data_range=(-420, 8848)):
+    '''
+    MinMaxScaler for elevations on Earth
+    Credit: @jhamman
+    '''
+    fmin, fmax = feature_range
+    dmin, dmax = data_range
+    scale = (fmax - fmin) / (dmax - dmin)
+    x_scaled = scale * x + fmin - dmin * scale
+    return x_scaled
+
+def cube(x):
+    """
+    Taking a cube, for the default inverse transformers
+    For some reason numpy doesn't like keyword arguments,
+    so you can'd do np.power(x, x2=3). This is necessary
+    to get around that, so that we can specify it as the
+    ``inverse_func``.
+    """
+    return np.power(x, 3)
+
+
+def fit_transformers(df: pd.DataFrame, custom_transformers: dict={}):
+    """
+    Fit function transformers for preprocessing data to train
+    a machine learning model. Note that this does not actually do
+    any data transformations, but merely fits the transformers
+    and returns them for later application. Any variables not in
+    the standard set provided here will simply be standardized
+
+    Parameters
+    ----------
+    df:
+        Dataframe with complete set of all data for training
+        This should have a MultiIndex with levels ('site', 'time')
+    custom_transformers:
+        Any custom transformers to be applied. These will override
+        the default ones specified here.
+
+    Returns
+    -------
+    fit_transformers:
+        Dictionary of transformers in the form {varname: fit_transformer}
+    """
+    # Default transformers
+    transformers = {
+        'raw_flow': Pipeline([
+            ('log1p', FunctionTransformer(func=np.log1p, inverse_func=np.expm1,
+                                          validate=False)),
+            ('normalize', MinMaxScaler())
+            ]),
+        'target_flow': Pipeline([
+            ('log1p', FunctionTransformer(func=np.log1p, inverse_func=np.expm1,
+                                          validate=False)),
+            ('normalize', MinMaxScaler())
+            ]),
+        'precipitation': Pipeline([
+            ('cbrt', FunctionTransformer(func=np.cbrt, inverse_func=cube,
+                                         validate=False)),
+            ('normalize', MinMaxScaler())
+            ]),
+        'elevation': FunctionTransformer(elevation_scaler),
```

Comment:
Why do you normalize elevation differently than other variables?

Reply:
Since we a priori know the range of elevations on Earth, we can just use them as the minimum and maximum values. That means that our transformer is not sensitive to biases in site selection.